elastic/apm-server
# spanmetrics: aggregate on span.name

## Motivation/summary

Add `span.name` to the service_destination metrics aggregation key. To protect against high cardinality span names (which will eventually be fixed, but are currently a problem in some instrumentation) we will stop aggregating on `span.name` once 50% of the aggregation groups capacity is reached, and log a warning.

## Checklist


## How to test these changes

1. Instrument an application to send 10000 exit spans to APM Server with low cardinality span names
2. Check that APM Server produces `service_destination` metrics all with `span.name` set
3. Update the instrumentation to use high cardinality span names, and send another 10000 exit spans such that the 10000 limit is reached within 1 minute.
4. Check that APM Server produces 5000 metrics with `span.name` set, and logs a warning once that 50% capacity is reached. The remaining metrics are captured without `span.name` set. (Because we only send 10000 events, there should be no immediately published metricsets.)

## Related issues

Closes #8274
